### ISSUE TITLE
Spatial constraints with Parcellations

### ISSUE THREAD

**Issue by dPys** [nilearn/nilearn#1750]
Parcellations is a great tool!

I'm wondering about how it handles spatial constraints-- when I use method='ward' and include a nifti mask object to constrain the clustering, how can I prevent getting solutions with spatially non-contiguous clusters?

Thanks,
Derek;

**Comment by GaelVaroquaux** [comment]
It should do it automatically. Doesn't it?

**Comment by dPys** [comment]
![ward_clust_nil](https://user-images.githubusercontent.com/16432683/44958551-5e51f100-aea7-11e8-99eb-fa937c8cb8d1.png)

This is an example solution using ward, where I've used a mask to constrain the clustering. Note that the two clusters circled in red have the same label generated by Parcellations, despite their obvious non-contiguity.

Please let me know if I'm just missing something obvious here!
-Derek

**Comment by GaelVaroquaux** [comment]
Did the mask that you used have several connect components? If you do something like this, the notion of connectivity in ward clustering is broken, and we cannot ensure that each cluster spans only one connected component. You need to apply parcellation separately per connected component.

**Comment by dPys** [comment]
It does seem that the mask has several connected components indeed. See attached img. 
[triple_net_ICA_overlap_3_sig_bin.nii.gz](https://github.com/nilearn/nilearn/files/2343610/triple_net_ICA_overlap_3_sig_bin.nii.gz)

What would be the easiest way to isolate connected components in the mask and apply Parcellations to each? I should also mention that when I use spectral NCUT with this mask, it seems to take care of this problem automatically, so I'm guessing the connected components assumption is specific to Ward/hierarchical methods?

**Comment by bthirion** [comment]
You may want to consider:
https://nilearn.github.io/modules/generated/nilearn.regions.connected_label_regions.html
Apply Ward to each resulting image.

And yes, the spatial constraints are specific to Ward.

**Comment by dPys** [comment]
FYI: Working on the latter method. Will get back to you guys in ~ a week or so to let you know how it goes (if you don't mind keeping this issue listed as open)

Thanks for the help :)

**Comment by KamalakerDadi** [comment]
Is there anyway that we could do an example out of this process ?

**Comment by dPys** [comment]
Hi @KamalakerDadi , Sorry for the very delayed reply. I was just going back through my old posts and saw that you had replied to this.  Would be happy to work out an example for this, but would like your feedback first on the best approach.

Here's what I have so far:
```
from nilearn import plotting
from nilearn.regions import connected_label_regions, parcellations

clust_mask = '/Users/PSYC-dap3463/Downloads/triple_net_ICA_overlap_3_sig_bin.nii.gz'
func_file = '/Users/PSYC-dap3463/Downloads/002/002.nii.gz'

def gen_img_list(dict_nii_img):
    from nilearn.image import new_img_like
    bna_img = nib.load(dict_nii_img)
    bna_data = np.round(bna_img.get_data(), 1)
    # Get an array of unique parcels
    bna_data_for_coords_uniq = np.unique(bna_data)
    # Number of parcels:
    par_max = len(bna_data_for_coords_uniq) - 1
    bna_data = bna_data.astype('int16')
    img_stack = []
    for idx in range(1, par_max+1):
        roi_img = bna_data == bna_data_for_coords_uniq[idx].astype('int16')
        roi_img = roi_img.astype('int16')
        img_stack.append(roi_img)
    img_stack = np.array(img_stack).astype('int16')
    img_list = []
    for idy in range(par_max):
        roi_img_nifti = new_img_like(bna_img, img_stack[idy])
        img_list.append(roi_img_nifti)
    bna_img.uncache()
    del img_stack
    return img_list

# Split binarized mask into connected component dictionary image    
conn_comps = connected_label_regions(clust_mask, min_size=50, connect_diag=True)

# Plot dictionary image
plotting.plot_roi(conn_comps, title='Image Dictionary of Connected Component Cluster Masks',
                  cut_coords=(8, -4, 9), colorbar=True, cmap='Paired')
                  
# Save dictionary image for reference
nib.save(conn_comps, '/Users/PSYC-dap3463/Downloads/connected_comps.nii.gz')

# Generate a list of connected component cluster masks using the dictionary from the previous steps
connect_comp_masks = gen_img_list('/Users/PSYC-dap3463/Downloads/connected_comps.nii.gz')

# Iteratively apply ward to each connected comp
ward_parc_img_list = []
for mask_img in connect_comp_masks:
    clust_est = Parcellations(method='ward', mask=mask_img, n_parcels=X)
    clust_est.fit(func_img)
    ward_parc_img_list.append(clust_est.labels_img_)
```

The trouble is that we still lack a heuristic for determining how many parcels to be extracted from each mask. If we wish to generate k=100 parcels, for instance, and we have 45 connected label regions from our mask, then what would be a preferable method for dividing up the 100 target parcels across the 45 separate masks? One thought would be to extract some subset number of parcels k_i from each mask proportionally to the number of voxels in each mask such that <a href="https://www.codecogs.com/eqnedit.php?latex=\sum_{i=1}^{nconn}k_{i}&space;=&space;k_{total}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\sum_{i=1}^{nconn}k_{i}&space;=&space;k_{total}" title="\sum_{i=1}^{nconn}k_{i} = k_{total}" /></a> where in my example nconn=45 and Ktotal=100

Curious to get your thoughts.

-Derek

**Comment by KamalakerDadi** [comment]
Hi @dPys Thanks for getting back.

I am not really sure about the example after thinking about it a bit (sorry). Because, spatial constraints are generated at no cost when you use Ward. But, your approach seems to me like not commonly used. Correct me if I am wrong.

Lets ask @GaelVaroquaux @bthirion what they think of an example.

**Comment by dPys** [comment]
Hi @KamalakerDadi , 

Although it is true that spatial constraints are generated at no cost with Ward, I do think that the scenario that I am describing here is actually fairly common-- i.e. one is attempting to parcellate using a cluster mask that includes multiple connected components. I think Parcellations may even benefit from an additional option to automatically: 1) break up the mask into all separate connected components; 2) apply the Parcellation method to each component separately (and even in parallel quite frankly); and 3) combine and reindex the resulting parcellations to yield a single atlas.

I can start working on a pull request for this or an example, if you're on board. And if not, happy to just work on this outside of the context of Nilearn as well.

Cheers,
@dPys

**Comment by GaelVaroquaux** [comment]
Hi @dPys,

When the mask has several connect components, hierarchical clustering is not well defined from a mathematical standpoint. In the strategy that you are describe, where it shows up is: what should be the number of clusters on each connect components? The Ward strategy that we use cannot give an answer to this question. How were you planning to tackle it?

**Comment by dPys** [comment]
Hi @GaelVaroquaux ,

Why not just pick a really simple heuristic like Hagenbach-Bischoff quota? this amounts to proportionally dividing up the total clusters intended into cluster 'bins' relative to the number of voxels in each connected component? 

Here's an example of how this could work (there may also be better ways of doing the atlas combinations using existing nilearn functions rather than what I do here):
```
import numpy as np
import nibabel as nib
from nilearn.regions import connected_regions
from nilearn.image import iter_img
from nilearn.regions import Parcellations
from nilearn.regions import connected_label_regions

# User specified number of clusters
k = 150

# Clustering method
clust_type = 'ward'

# Mask file with multiple connected components
mask_file = '/Users/PSYC-dap3463/Downloads/triple_net_ICA_overlap_3_sig_bin.nii.gz'

# EPI Functional file
func_file = '/Users/PSYC-dap3463/Downloads/002/002.nii.gz'

# Output file path for super-atlas
out_path = '/Users/PSYC-dap3463/Downloads/002/super_atlas.nii.gz'


def proportional(k, voxels_list):
    '''Hagenbach-Bischoff Quota'''
    quota=sum(voxels_list)/(1.+k)
    frac=[voxels/quota for voxels in voxels_list]
    res=[int(f) for f in frac]
    n=k-sum(res)
    if n==0: return res
    if n<0: return [min(x,k) for x in res]
    remainders=[ai-bi for ai,bi in zip(frac,res)]
    limit=sorted(remainders,reverse=True)[n-1]
    for i,r in enumerate(remainders):
        if r>=limit:
            res[i]+=1
            n-=1
            if n==0: return res
    raise

def gen_img_list(bna_img):
    from nilearn.image import new_img_like
    bna_data = np.round(bna_img.get_data(), 1)
    # Get an array of unique parcels
    bna_data_for_coords_uniq = np.unique(bna_data)
    # Number of parcels:
    par_max = len(bna_data_for_coords_uniq) - 1
    bna_data = bna_data.astype('int16')
    img_stack = []
    for idx in range(1, par_max+1):
        roi_img = bna_data == bna_data_for_coords_uniq[idx].astype('int16')
        roi_img = roi_img.astype('int16')
        img_stack.append(roi_img)
    img_stack = np.array(img_stack).astype('int16')
    img_list = []
    for idy in range(par_max):
        roi_img_nifti = new_img_like(bna_img, img_stack[idy])
        img_list.append(roi_img_nifti)
    bna_img.uncache()
    del img_stack
    return img_list

def flatten(l):
    import collections
    for el in l:
        if isinstance(el, collections.Iterable) and not isinstance(el, (str, bytes)):
            for ell in flatten(el):
                yield ell
        else:
            yield el

def create_parcel_atlas(parcel_list):
    from nilearn.image import new_img_like, concat_imgs
    parcel_background = new_img_like(parcel_list[0], np.zeros(parcel_list[0].shape, dtype=bool))
    parcel_list_exp = [parcel_background] + parcel_list
    parcellation = concat_imgs(parcel_list_exp).get_data()
    index_vec = np.array(range(len(parcel_list_exp))) + 1
    net_parcels_sum = np.sum(index_vec * parcellation, axis=3)
    net_parcels_map_nifti = nib.Nifti1Image(net_parcels_sum, affine=parcel_list[0].affine)
    return net_parcels_map_nifti, parcel_list_exp


func_img = nib.load(func_file)

[conn_comps_img, _]= connected_regions(mask_file, extract_type='connected_components')

conn_comp_list = iter_img(conn_comps_img)

mask_img_list = [] 
mask_voxels_dict = dict()
for i, mask_img in enumerate(iter_img(conn_comp_list)):
    mask_voxels_dict[i] = np.int(np.sum(mask_img.get_data()))
    mask_img_list.append(mask_img)

k_list = proportional(k, list(mask_voxels_dict.values()))

conn_comp_atlases = []
i=0
for mask_img in mask_img_list:
    clust_est = Parcellations(method=clust_type, n_parcels=k_list[i], mask=mask_img)
    clust_est.fit(func_img)
    region_labels = connected_label_regions(clust_est.labels_img_)
    conn_comp_atlases.append(region_labels)
    i = i + 1
                            
# Then combine the multiple atlases, corresponding to each connected component, into a single atlas
atlas_of_atlases = []
for atlas in conn_comp_atlases:
    atlas_of_atlases.append(gen_img_list(atlas))

atlas_of_atlases = list(flatten(atlas_of_atlases))

[super_atlas_ward, _] = create_parcel_atlas(atlas_of_atlases)

nib.save(super_atlas_ward, out_path)
```

Super atlas:
<img width="1237" alt="screen shot 2018-12-01 at 1 58 23 am" src="https://user-images.githubusercontent.com/16432683/49325757-3f4a1600-f50d-11e8-9460-a73e4d32f856.png">

@dPys

**Comment by GaelVaroquaux** [comment]
I wrote a long reply below, trying to think it over. But when I reread the whole conversation of this thread, it all boils down to the fact that in the case of multiple connect components in the mask, the method and the users expectations are ill defined. Ward is a variance-minimizing clustering algorithm, forcing it to respect connect components boundary is hard without completely breaking its variance-minimizing properties. Breaking them would then create behaviors that break the expectations of users (maybe not the same users :D).

> Why not just pick a really simple heuristic like Hagenbach-Bischoff quota? this amounts to proportionally dividing up the total clusters intended into cluster 'bins' relative to the number of voxels in each connected component?

Such a heuristic would not be the Ward algorithm, as it would not be compatible with the Ward objective of minimizing the variance. The reason that the Ward algorithm works well is specifically this variance minimizing objective, which controls the quality of fit to the data. My experience is that quick hacks are seldom well behaved.

To come back to your heuristic: if a connected components contains only one constant value, it should not be broken into multiple clusters. What if two connected components contain the same value, should we break them up, or not? From a variance-minimizing standard, we should merge them. For certain users, we should not. How do we ensure that for two masks very similar, one with one connected component, the other with two, the results are close? 

Now that said, I think that the Ward criterion could be adapted to work jointly in multiple connect components without merging them. We could apply it separately to each connected component, get a tree in each of them, and find the cut with the specified amount of clusters that minimizes the error in variance. This would optimize the number of cluster in each connected component to minimize the error in variance. It would be a bit of work, and doing it efficiently would be even more work (the naive implementation that I can think of would do many comparison, but I think that the distances output by the Ward tree could be used to reduce the cost).

If the above proposal can be done reasonably efficiently (ie without an n**2 cost), it would be a fantastic addition to nilearn.

**Comment by dPys** [comment]
Hi @GaelVaroquaux 

Thank you for the very thoughtful response and suggestions. I think this:
_"How do we ensure that for two very similar masks, one with one connected component, the other with two, the results are close?"_
captures the problem well if we neglect variance-minimization in the heuristic. And it's fair to say that that could present issues for certain user cases.

As for the challenge you presented, I can try. Since as you noted it would involve quite a bit of work to come up with something efficient, it might be awhile before I have a good prototype. This would also mean going back to working directly with Ward from sklearn and then reintegrating with Parcellations. But given since I think that users likely encounter this scenario not-too-infrequently, it might be a worthwhile investment? 

Will tinker, stay tuned.

@dPys

**Comment by rickwith** [comment]
Actually i am trying to do this spatial constraints thing on fmri data, it would be a great addition to nilearn package.

**Comment by bthirion** [comment]
See  https://github.com/scikit-learn/scikit-learn/pull/18964
